rfcs/0000-compiled-runtime.md

```diff
@@ -0,0 +1,533 @@
+
+-const blackStyles = {
++const blackStyles = css({
+  color: 'black'
+});
+
+function Component({ isBlack }) {
+  return <div css={isBlack && blackStyles} />;
+}
+```
+
+Compiled would:
+
+1. Find the `css` function call
+2. Generate class name and styles
+3. Lean on the runtime to do the rest
+
+You can see a basic implementation of the above on the [`babel-experiment` branch](https://github.com/atlassian-labs/compiled/compare/babel-experiment?expand=1#diff-7c3b8f72b34155cdf17c42f282ed0cecec4fd2863ba71912213f8932159cd01bR17). Note only tests work, Storybook is broken.
+
+A more technical design overview can be found further into the RFC.
+
+## Motivation
+
+The primary motivation is for Compiled maintainers to maintain and write less code to handle the ever increasing list of edge cases and consumers having a better developer experience because things JustWork with any number of dynamic application of styles, with supplementary motivators being improved build performance as there is less AST traversal and transformation required at build time.
+
+### Technical design considerations
+
+- 🤏 Minimize AST traversal & transformation
+- ⛓️ One nodes transformation should not be dependent on another nodes
+- 🥇 Ignore intermediate nodes - let the runtime handle it!
+
+## Detail design
+
+This change should work with all Compiled APIs and ensure style extraction is backwards compatible for at least one major version.
+
+### CSS
+
+Style rules now must be defined using clear call sites using `css` instead of raw objects or strings. We can still leverage the large majority of code already written to statically evaluate styles. The main difference however is instead of being opaque and resolving to `null` they now get transformed into a data structure Compiled can understand at runtime.
+
+```jsx
+import { css } from '@compiled/react';
+
+const style = css({
+  color: 'black',
+  ':hover': {
+    color: 'red',
+  },
+});
+
+// 👇 transforms to
+
+import { i } from '@compiled/react/insert';
+
+const style = i([
+  '_syaz11x8 _30l35scu',
+  ['._syaz11x8{color:black}', '._30l35scu:hover{color:red}'],
+]);
+```
+
+An identity function is used to flag call sites for extraction later (naming TBD). If the style rules are not declared at the top of the module they will be hoisted if safe to the top of the module.
+
+```jsx
+/** @jsxImportSource @compiled/react */
+
+function Comp() {
+  return <div css={css({ color: 'black' })} />;
+}
+
+// 👇 transforms to
+
+const _1 = i(['_syaz11x8', ['._syaz11x8{color:black}']]);
+
+function Comp() {
+  return <div css={_1} />;
+}
+```
+
+When referencing an expression Compiled would try to statically evaluate it else if it fails be added as the third item in the array along with all other dynamic styles.
+
+```jsx
+import { N800 } from '@atlaskit/theme/colors';
+
+const styles = css({ color: N800 });
+
+// 👇 transforms to
+
+const styles = i(['_syaz4rde', ['._syaz4rde{color: var(--_kmurgp)}'], [['--_kmurgp', N800]]]);
```

Comment:
How might styles with suffix/prefix' affect dynamic styles? Lift and shift should work fine.

```jsx
import { css } from '@compiled/react';
import image from '!url-loader!./image.png';

const styles = css({ background: `url(${image})` });
```

```jsx
const styles = i(['_1asd931s', ['._1asd931s{background: var(--_921awg)}'], [['--_921awg', `url(${image})`]]]);
```